sol-transactions/generate.go

```diff
@@ -0,0 +1,223 @@
+package soltransactions
+
+import (
+	"bytes"
+	"context"
+	"embed"
+	"fmt"
+	"os"
+	"time"
+
+	"strings"
+
+	"github.com/streamingfast/dgrpc"
+	codegen "github.com/streamingfast/substreams-codegen"
+	"github.com/streamingfast/substreams-codegen/loop"
+	pbbuild "github.com/streamingfast/substreams-codegen/pb/sf/codegen/remotebuild/v1"
+	"go.uber.org/zap"
+)
+
+//go:embed templates/*
+var templatesFS embed.FS
+
+func cmdGenerate(p *Project) loop.Cmd {
+	return func() loop.Msg {
+		projFiles, err := p.generate()
+		if err != nil {
+			return codegen.ReturnGenerate{Err: err}
+		}
+		return codegen.ReturnGenerate{
+			ProjectFiles: projFiles,
+		}
+	}
+}
+
+func cmdBuild(p *Project) loop.Cmd {
+	p.buildStarted = time.Now()
+	p.compilingBuild = true
+
+	return func() loop.Msg {
+		buildResponseChan := make(chan *codegen.RemoteBuildState, 1)
+		go func() {
+			p.build(buildResponseChan)
+			close(buildResponseChan)
+		}()
+
+		// go to the state of compiling build
+		return codegen.CompilingBuild{
+			FirstTime:       true,
+			RemoteBuildChan: buildResponseChan,
+		}
+	}
+}
+func cmdBuildFailed(logs []string, err error) loop.Cmd {
+	return func() loop.Msg {
+		return codegen.ReturnBuild{Err: err, Logs: strings.Join(logs, "\n")}
+	}
+}
+
+func cmdBuildCompleted(content *codegen.RemoteBuildState) loop.Cmd {
+	return func() loop.Msg {
+		return codegen.ReturnBuild{
+			Err:       nil,
+			Logs:      strings.Join(content.Logs, "\n"),
+			Artifacts: content.Artifacts,
+		}
+	}
+}
+
+func (p *Project) generate() (projFiles map[string][]byte, err error) {
+	// TODO: before doing any generation, we'll want to validate
+	// all data points that are going into source code.
+	// We don't want some weird things getting into `build.rs`
+	// and being executed server side, so we'll need pristine validation
+	// of all inputs here.
+	// TODO: add some checking to make sure `ParentContractName` of DynamicContract
+	// do match a Contract that exists here.
+
+	projFiles, err = p.Render()
+	if err != nil {
+		return nil, fmt.Errorf("rendering template: %w", err)
+	}
+
+	return
+}
+
+func (p *Project) build(remoteBuildContentChan chan<- *codegen.RemoteBuildState) {
```

Comment:
lots of duplications for remote building here.. let's purge, in any case, when we re-enable, it's most certainly going to be abstracted away, not duplicated 7 times like right now